src/borda_clj/client.clj

```diff
@@ -0,0 +1,82 @@
+(ns borda-clj.client
+  (:use [slingshot.slingshot :only [throw+ try+]])
+  (:require [aleph.http :as http]
+            [byte-streams :as bs]
+            [cheshire.core :as json]))
+
+(def discard_key {:op "_discard"})
+
+(defn empty-measurements []
+  {discard_key {:success_count 0}})
+
+(defn collect
+  "Adds the given measurement (values and dimensions) to the given hashmap and caps the size of the hashmap at the given max-buffer-size"
+  [measurements max-buffer-size dimensions values]
+  (let [key (into (sorted-map) dimensions)]
+    (if (contains? measurements key)
+      ; TODO: support stuff other than SUM here (e.g. AVG, MIN, MAX, etc.)
+      (update measurements key (partial merge-with +) values) ; merge with existing in buffer
+      (if (< (count measurements) max-buffer-size)
+        (assoc measurements key values) ; space available, add to buffer
+        (do (print "borda buffer full, discarding measurement" dimensions values)
+            (update measurements discard_key (partial merge-with +) {:success_count 1})))))) ; buffer full
+
+(defn merge-global [global-dimensions kv]
+  (let [[dimensions values] kv]
+    [(merge global-dimensions dimensions) values]))
+
+(defn merge-global-to-measurements [global-dimensions measurements]
+  (into {} (map (partial merge-global global-dimensions) (seq measurements))))
+
+(defn reducing-submitter
+  "Returns two functions. The first is a reducing submitter that collects and
+   aggregates measurements and sends them using the given send function at the
+   specified interval (in milliseconds from end of one send to the beginning of
+   the next send) and limits the number of pending measurements to
+   max-buffer-size. The second is a function that flushes the buffer and stops
+   the background thread that does the submitting. Parameter on-send-error is a
+   function that will be called with a map of buffered measurements and the
+   error if and when flushing the buffer to borda fails. Except on the final
+   flush, failed measurements will be rebuffered if space is available.
+   Parameter global-dimensions is a map of dimensions that will be included with
+   every single measurement."
+  [global-dimensions max-buffer-size interval send on-send-error]
```

Comment:
@aranhoide I just added this `global-dimensions` concept. It allows us to capture global stuff like the app name, server hostname, etc by passing it once here, and then it automatically gets applied to all measurements. Besides being convenient, it also gives us a way to associate dimensions with the auto-generated `_discard` op so that we'll be able to tell which app/client/etc is doing all the discarding.